tests/tizen_capi/unittest_tizen_capi_latency.cc

```diff
@@ -0,0 +1,322 @@
+      EXPECT_EQ (status, ML_ERROR_NONE);
+
+      /** Allocate input data */
+      input = output = NULL;
+      status = ml_tensors_data_create (in_info, &input);
+      EXPECT_EQ (status, ML_ERROR_NONE);
+      EXPECT_TRUE (input != NULL);
+      ml_tensors_info_destroy (in_info);
+
+      /** Load input data into the buffer */
+      status =
+        ml_tensors_data_get_tensor_data (input, 0, (void **) &data, &data_size);
+      EXPECT_EQ (status, ML_ERROR_NONE);
+      if (fd >= 0) {
+        resetDataFile();
+        data_read = read(fd, data, data_size);
+        EXPECT_EQ (data_read, data_size);
+      }
+
+      /** Initial run to warm up the cache */
+      status = ml_single_invoke (single, input, &output);
+      EXPECT_EQ (status, ML_ERROR_NONE);
+      EXPECT_TRUE (output != NULL);
+      status =
+        ml_tensors_data_get_tensor_data (output, 0, (void **) &data, &data_size);
+      EXPECT_EQ (status, ML_ERROR_NONE);
+      matchOutput (data, data_size);
+      ml_tensors_data_destroy (output);
+
+      /** Benchmark the invoke duration */
+      for (int idx=0; idx<RUN_COUNT; ++idx) {
+        start = g_get_real_time();
+        status = ml_single_invoke (single, input, &output);
+        end = g_get_real_time();
+        single_invoke_duration += end - start;
+        EXPECT_EQ (status, ML_ERROR_NONE);
+        EXPECT_TRUE (output != NULL);
+        ml_tensors_data_destroy (output);
+      }
+
+      single_invoke_duration_f = (single_invoke_duration * 1.0)/RUN_COUNT;
+      g_warning ("Time to invoke single = %f us", single_invoke_duration_f);
+
+      /** Close the single handle */
+      status = ml_single_close (single);
+      EXPECT_EQ (status, ML_ERROR_NONE);
+
+      ml_tensors_data_destroy (input);
+    }
+
+    /** Benchmark the invoke time for the filter directly */
+    void benchmarkDirectInvoke(const char *filter) {
+      GstTensorMemory input, output;
+      GstTensorsInfo res;
+
+      const gchar *model_files[] = { model_file, NULL, };
+      GstTensorFilterProperties prop = {
+        .fwname = filter,
+        .fw_opened = 0,
+        .model_files = model_files,
+        .num_models = 1,
+      };
+
+      /** Find the filter */
+      const GstTensorFilterFramework *sp = nnstreamer_filter_find (filter);
+      EXPECT_NE (sp, (void *) NULL);
+
+      /** Open the filter */
+      status = sp->open (&prop, &data);
+      EXPECT_EQ (status, 0);
+
+      /** get input/output data info */
+      status = sp->getInputDimension (&prop, &data, &res);
+      EXPECT_EQ (status, 0);
+
+      input.type = res.info[0].type;
+      input.size = gst_tensor_info_get_size (&res.info[0]);
+
+      status = sp->getOutputDimension (&prop, &data, &res);
+      EXPECT_EQ (status, 0);
+
+      output.type = res.info[0].type;
+      output.size = gst_tensor_info_get_size (&res.info[0]);
+
+      /** Allocate input and output data */
+      input.data = g_malloc(input.size);
+      output.data = g_malloc(output.size);
```

Comment:
@myungjoo This is one of the key differences between single and direct invoke. Single API invoke allocates memory for output internally with each call to invoke. So, a direct comparison to invoke between single and direct add extra overhead of this output data allocation. 
Should this allocation be moved to time critical section for a fair comparison?

Reply:
If it is malloc'ed by app, it is not our problem (not our latency). If it is malloc'ed by nnstreamer's code (not app or unittest), it is our problem (our latency)